File: android/app/src/main/java/com/slackoff/app/playback/PlaybackEngine.kt
package com.slackoff.app.playback

import android.app.Application
import androidx.media3.common.MediaItem
import androidx.media3.common.PlaybackException
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

enum class PlaybackMode {
    IDLE, SLIDESHOW, EXOPLAYER
}

class PlaybackEngine(private val application: Application) {
    private var _mode = MutableStateFlow(PlaybackMode.IDLE)
    val mode: StateFlow<PlaybackMode> = _mode.asStateFlow()

    private var _segments = MutableStateFlow<List<M3u8Segment>>(emptyList())
    val segments: StateFlow<List<M3u8Segment>> = _segments.asStateFlow()

    private var _currentIndex = MutableStateFlow(0)
    val currentIndex: StateFlow<Int> = _currentIndex.asStateFlow()

    private var _elapsed = MutableStateFlow(0.0)
    val elapsed: StateFlow<Double> = _elapsed.asStateFlow()

    private var _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private var _isBuffering = MutableStateFlow(false)
    val isBuffering: StateFlow<Boolean> = _isBuffering.asStateFlow()

    private var _isPlaying = MutableStateFlow(false)
    val isPlaying: StateFlow<Boolean> = _isPlaying.asStateFlow()

    private var _speed = MutableStateFlow(1.0)
    val speed: StateFlow<Double> = _speed.asStateFlow()

    private var _volume = MutableStateFlow(1.0f)
    val volume: StateFlow<Float> = _volume.asStateFlow()

    private var _isMuted = MutableStateFlow(false)
    val isMuted: StateFlow<Boolean> = _isMuted.asStateFlow()

    private var _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private var _playerDuration = MutableStateFlow(0.0)
    val playerDuration: StateFlow<Double> = _playerDuration.asStateFlow()

    private var _playerTime = MutableStateFlow(0.0)
    val playerTime: StateFlow<Double> = _playerTime.asStateFlow()

    /** 供 UI 观察的当前播放时间（tick / seek 时刷新）。 */
    private val _currentTimeFlow = MutableStateFlow(0.0)
    val currentTimeFlow: StateFlow<Double> = _currentTimeFlow.asStateFlow()

    private fun publishTime() {
        _currentTimeFlow.value = currentTime
    }

    var exoPlayer: ExoPlayer? = null
        private set

    private var source: HlsMediaSource? = null
    private var tickJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    val currentSegment: M3u8Segment?
        get() {
            val segs = _segments.value
            val idx = _currentIndex.value
            return if (idx in segs.indices) segs[idx] else null
        }

    val playlistDuration: Double
        get() = _segments.value.sumOf { it.duration }

    val totalDuration: Double
        get() {
            if (_mode.value == PlaybackMode.EXOPLAYER && _playerDuration.value > 0) {
                return _playerDuration.value
            }
            return playlistDuration
        }

    val currentTime: Double
        get() {
            if (_mode.value == PlaybackMode.EXOPLAYER) return _playerTime.value
            val segs = _segments.value
            val idx = _currentIndex.value
            val previous = segs.take(idx).sumOf { it.duration }
            return previous + minOf(_elapsed.value, segs.getOrNull(idx)?.duration ?: 0.0)
        }

    val progress: Float
        get() {
            val dur = totalDuration
            if (dur <= 0) return 0f
            return (currentTime / dur).toFloat().coerceIn(0f, 1f)
        }

    fun setSpeed(speed: Double) {
        _speed.value = speed
        exoPlayer?.setPlaybackSpeed(speed.toFloat())
    }

    fun setVolume(volume: Float) {
        _volume.value = volume
        if (!_isMuted.value) {
            exoPlayer?.volume = volume
        }
    }

    fun setMuted(muted: Boolean) {
        _isMuted.value = muted
        exoPlayer?.volume = if (muted) 0f else _volume.value
    }

    suspend fun load(streamUrl: String, headers: Map<String, String> = emptyMap()) {
        stopTicking()
        teardown()
        _isLoading.value = true
        _errorMessage.value = null

        try {
            // 判断是否是 MP4 直链（不是 M3U8 播放列表）
            val isDirectMp4 = isDirectVideoUrl(streamUrl)

            if (isDirectMp4) {
                // MP4 直链：直接用 ExoPlayer 播放，不需要 HLS 本地代理
                startDirectVideo(streamUrl, headers)
            } else {
                // M3U8 播放列表：走 HLS 本地代理流程
                val playlistData = withContext(Dispatchers.IO) {
                    MediaFetcher.fetch(streamUrl, headers)
                }
                if (playlistData == null) {
                    _errorMessage.value = "无法加载播放列表"
                    _isLoading.value = false
                    return
                }
                val text = String(playlistData, Charsets.UTF_8)
                val parsed = M3u8Parser.parse(text, streamUrl)
                _segments.value = parsed
                _currentIndex.value = 0
                _elapsed.value = 0.0

                val kind = withContext(Dispatchers.IO) {
                    probe(parsed, headers)
                }

                if (kind == SegmentKind.IMAGE) {
                    _mode.value = PlaybackMode.SLIDESHOW
                } else {
                    startVideo(streamUrl, headers, playlistData, kind)
                }
            }
        } catch (e: Exception) {
            _errorMessage.value = e.message
            _mode.value = PlaybackMode.IDLE
        }

        _isLoading.value = false
        startTicking()
    }

    /** 判断是否是直接的视频文件（MP4 等），而不是 M3U8 播放列表 */
    private fun isDirectVideoUrl(url: String): Boolean {
        val path = url.substringBefore("?").substringBefore("#").lowercase()
        return path.endsWith(".mp4") ||
               path.endsWith(".m4v") ||
               path.endsWith(".mov") ||
               path.endsWith(".webm") ||
               path.endsWith(".mkv") ||
               path.endsWith(".avi")
    }

    /** 直接播放视频文件（MP4 等），不经过 HLS 本地代理 */
    private suspend fun startDirectVideo(url: String, headers: Map<String, String>) =
        withContext(Dispatchers.Main) {
            val player = ExoPlayer.Builder(application).build()
            player.volume = if (_isMuted.value) 0f else _volume.value
            player.setPlaybackSpeed(_speed.value.toFloat())

            player.addListener(object : Player.Listener {
                override fun onPlaybackStateChanged(state: Int) {
                    _isBuffering.value = state == Player.STATE_BUFFERING
                    _isPlaying.value = state == Player.STATE_READY && player.playWhenReady
                }

                override fun onPlayerError(error: PlaybackException) {
                    _errorMessage.value = error.message
                }
            })

            val mediaItem = MediaItem.fromUri(url)
            player.setMediaItem(mediaItem)
            player.prepare()
            player.playWhenReady = true

            exoPlayer = player
            _mode.value = PlaybackMode.EXOPLAYER
            _isPlaying.value = true
        }

    private suspend fun probe(segments: List<M3u8Segment>, headers: Map<String, String>): SegmentKind =
        withContext(Dispatchers.IO) {
            val first = segments.firstOrNull() ?: return@withContext SegmentKind.UNKNOWN
            val data = MediaFetcher.fetch(first.url, headers, 0 until 65536)
                ?: return@withContext SegmentKind.UNKNOWN
            TsSegmentRepair.kindOf(data)
        }

    private suspend fun startVideo(
        origin: String,
        headers: Map<String, String>,
        playlist: ByteArray,
        kind: SegmentKind
    ) = withContext(Dispatchers.Main) {
        val ext = if (kind == SegmentKind.FRAGMENTED_MP4) "m4s" else "ts"
        val (url, src) = HlsLocalServer.mount(
            origin = origin,
            headers = headers,
            segmentExtension = ext,
            playlist = playlist
        )
        source = src

        val player = ExoPlayer.Builder(application).build()
        player.volume = if (_isMuted.value) 0f else _volume.value
        player.setPlaybackSpeed(_speed.value.toFloat())

        player.addListener(object : Player.Listener {
            override fun onPlaybackStateChanged(state: Int) {
                _isBuffering.value = state == Player.STATE_BUFFERING
                _isPlaying.value = state == Player.STATE_READY && player.playWhenReady
            }

            override fun onPlayerError(error: PlaybackException) {
                _errorMessage.value = error.message
            }
        })

        val mediaItem = MediaItem.fromUri(url)
        player.setMediaItem(mediaItem)
        player.prepare()
        player.playWhenReady = true

        exoPlayer = player
        _mode.value = PlaybackMode.EXOPLAYER
        _isPlaying.value = true
    }

    fun play() {
        when (_mode.value) {
            PlaybackMode.SLIDESHOW -> {
                val segs = _segments.value
                val idx = _currentIndex.value
                if (idx == segs.size - 1 && _elapsed.value >= (segs.getOrNull(idx)?.duration ?: 0.0)) {
                    seekToFraction(0f)
                }
                _isPlaying.value = true
                startTicking()
            }
            PlaybackMode.EXOPLAYER -> {
                exoPlayer?.playWhenReady = true
                _isPlaying.value = true
            }
            PlaybackMode.IDLE -> {}
        }
    }

    fun pause() {
        _isPlaying.value = false
        exoPlayer?.playWhenReady = false
    }

    fun tick(delta: Double) {
        when (_mode.value) {
            PlaybackMode.SLIDESHOW -> {
                if (!_isPlaying.value) return
                val segs = _segments.value
                val idx = _currentIndex.value
                val segment = segs.getOrNull(idx) ?: return
                _elapsed.value = _elapsed.value + delta * maxOf(_speed.value, 0.25)
                if (_elapsed.value >= segment.duration) {
                    if (idx + 1 < segs.size) {
                        _currentIndex.value = idx + 1
                        _elapsed.value = 0.0
                    } else {
                        _elapsed.value = segment.duration
                        _isPlaying.value = false
                    }
                }
            }
            PlaybackMode.EXOPLAYER -> {
                val player = exoPlayer ?: return
                val seconds = player.currentPosition / 1000.0
                if (seconds.isFinite() && seconds >= 0) _playerTime.value = seconds
                val duration = player.duration / 1000.0
                if (duration.isFinite() && duration > 0) _playerDuration.value = duration
            }
            PlaybackMode.IDLE -> {}
        }
        publishTime()
    }

    fun seekToFraction(fraction: Float) {
        val clamped = fraction.coerceIn(0f, 1f)
        when (_mode.value) {
            PlaybackMode.SLIDESHOW -> {
                val segs = _segments.value
                if (segs.isEmpty()) return
                var remaining = clamped.toDouble() * totalDuration
                var index = segs.size - 1
                for ((i, seg) in segs.withIndex()) {
                    if (remaining <= seg.duration) {
                        index = i
                        break
                    }
                    remaining -= seg.duration
                }
                _currentIndex.value = index.coerceIn(0, segs.size - 1)
                _elapsed.value = remaining.coerceAtMost(segs[_currentIndex.value].duration)
            }
            PlaybackMode.EXOPLAYER -> {
                val dur = totalDuration
                if (dur > 0) {
                    exoPlayer?.seekTo((clamped * dur * 1000).toLong())
                }
            }
            PlaybackMode.IDLE -> {}
        }
        publishTime()
    }

    fun seekTo(time: Double) {
        if (totalDuration > 0) {
            seekToFraction((time / totalDuration).toFloat())
        }
    }

    fun startTicking() {
        tickJob?.cancel()
        tickJob = scope.launch {
            while (isActive) {
                delay(50)
                tick(0.05)
            }
        }
    }

    private fun stopTicking() {
        tickJob?.cancel()
        tickJob = null
    }

    fun stop() {
        stopTicking()
        teardown()
    }

    private fun teardown() {
        exoPlayer?.stop()
        exoPlayer?.release()
        exoPlayer = null
        source?.let { HlsLocalServer.getInstance().unregister(it) }
        source = null
        _mode.value = PlaybackMode.IDLE
        _segments.value = emptyList()
        _currentIndex.value = 0
        _elapsed.value = 0.0
        _isPlaying.value = false
        _isBuffering.value = false
        _playerTime.value = 0.0
        _playerDuration.value = 0.0
        _currentTimeFlow.value = 0.0
    }
}

File: SlackOff/Playback/PlaybackEngine.swift
import AVFoundation
import Foundation
import Observation

/// 播放引擎。
///
/// 视频走系统播放器：真实 m3u8 交给 `HLSLocalServer` 挂到本地回环地址上，
/// 由它下载分片、绕过图床防盗链、把「伪装成图片的 TS」修好再喂给 AVPlayer。
/// MP4 单文件源不走这套代理，发现是 mp4 就直接交给 AVPlayer 播。
/// 只有确认分片真的是图片时，才退回图片轮播。
@Observable @MainActor
final class PlaybackEngine {
    enum Mode: Equatable {
        case idle
        case slideshow
        case avplayer
    }

    private(set) var mode: Mode = .idle
    private(set) var segments: [M3U8Segment] = []
    private(set) var currentIndex = 0
    private(set) var elapsed: Double = 0
    private(set) var isLoading = false
    private(set) var isBuffering = false
    private(set) var player: AVPlayer?

    var isPlaying = false
    var speed: Double = 1.0 {
        didSet { applyRate() }
    }
    /// 播放器音量（0...1），手势右滑上下调节的就是它。
    var volume: Float = 1.0 {
        didSet { player?.volume = volume }
    }
    /// 静音开关。
    var isMuted = false {
        didSet { player?.isMuted = isMuted }
    }
    var errorMessage: String?

    private var tickTask: Task<Void, Never>?
    private var source: HLSMediaSource?
    private var playerTime: Double = 0
    private var playerDuration: Double = 0

    var currentSegment: M3U8Segment? {
        segments.indices.contains(currentIndex) ? segments[currentIndex] : nil
    }

    /// 播放列表里 EXTINF 累加出来的时长，真实时长拿到之前先用它顶着。
    var playlistDuration: Double {
        segments.reduce(0) { $0 + $1.duration }
    }

    var totalDuration: Double {
        if mode == .avplayer, playerDuration > 0 { return playerDuration }
        return playlistDuration
    }

    var currentTime: Double {
        if mode == .avplayer { return playerTime }
        let previous = segments.prefix(currentIndex).reduce(0) { $0 + $1.duration }
        return previous + min(elapsed, currentSegment?.duration ?? 0)
    }

    var progress: Double {
        totalDuration > 0 ? min(max(currentTime / totalDuration, 0), 1) : 0
    }

    // MARK: - 加载

    func load(streamURL url: URL, headers: [String: String] = [:], mtype: String = "m3u8") async {
        stopTicking()
        teardown()
        isLoading = true
        errorMessage = nil

        let isMP4 = mtype == "mp4" || url.pathExtension.lowercased() == "mp4"

        if isMP4 {
            startMP4(origin: url)
            isLoading = false
            startTicking()
            return
        }

        do {
            let playlistData = try await MediaFetcher.data(from: url, extraHeaders: headers)
            let text = String(data: playlistData, encoding: .utf8) ?? ""
            let parsed = M3U8Parser.parse(text, baseURL: url)
            reset()
            segments = parsed

            let kind = await Self.probe(parsed, headers: headers)
            if kind == .image {
                mode = .slideshow
            } else {
                try await startVideo(origin: url, headers: headers, playlist: playlistData, kind: kind)
            }
        } catch {
            reset()
            errorMessage = error.localizedDescription
        }

        isLoading = false
        startTicking()
    }

    /// 只取分片头部 64KB 判断真实类型：图片 / TS / 伪装成图片的 TS / fMP4。
    private static func probe(_ segments: [M3U8Segment], headers: [String: String]) async -> SegmentKind {
        guard let first = segments.first,
              let data = try? await MediaFetcher.data(from: first.url, extraHeaders: headers, range: 0 ..< 65_536) else {
            return .unknown
        }
        return TSSegmentRepair.kind(of: data)
    }

    private func startVideo(
        origin: URL,
        headers: [String: String],
        playlist: Data,
        kind: SegmentKind
    ) async throws {
        let (url, source) = try await HLSLocalServer.mount(
            origin: origin,
            headers: headers,
            segmentExtension: kind == .fragmentedMP4 ? "m4s" : "ts",
            playlist: playlist
        )
        self.source = source

        let player = AVPlayer(playerItem: AVPlayerItem(asset: AVURLAsset(url: url)))
        player.automaticallyWaitsToMinimizeStalling = true
        player.volume = volume
        player.isMuted = isMuted
        self.player = player
        mode = .avplayer
        isPlaying = true
        configureAudioSession()
        applyRate()
    }

    /// MP4 单文件播放：不走 m3u8 那套本地代理逻辑，AVPlayer 直接播远程地址。
    private func startMP4(origin: URL) {
        let player = AVPlayer(playerItem: AVPlayerItem(asset: AVURLAsset(url: origin)))
        player.automaticallyWaitsToMinimizeStalling = true
        player.volume = volume
        player.isMuted = isMuted
        self.player = player
        mode = .avplayer
        isPlaying = true
        configureAudioSession()
        applyRate()
    }

    // 播放状态（status / duration / error / 缓冲）统一在 tick 里读，省掉一套 KVO。

    // MARK: - 播控

    func play() {
        switch mode {
        case .slideshow:
            if currentIndex == segments.count - 1, elapsed >= (currentSegment?.duration ?? 0) {
                seek(toFraction: 0)
            }
            isPlaying = true
            startTicking()
        case .avplayer:
            isPlaying = true
            applyRate()
        case .idle:
            break
        }
    }

    func pause() {
        isPlaying = false
        if mode == .avplayer { player?.pause() }
    }

    func tick(delta: Double) {
        switch mode {
        case .slideshow:
            guard isPlaying, let segment = currentSegment else { return }
            elapsed += delta * max(speed, 0.25)
            if elapsed >= segment.duration {
                if currentIndex + 1 < segments.count {
                    currentIndex += 1
                    elapsed = 0
                } else {
                    elapsed = segment.duration
                    isPlaying = false
                }
            }
        case .avplayer:
            guard let player else { return }
            let seconds = player.currentTime().seconds
            if seconds.isFinite, seconds >= 0 { playerTime = seconds }
            if let item = player.currentItem {
                let duration = item.duration.seconds
                if duration.isFinite, duration > 0 { playerDuration = duration }
                if item.status == .failed, errorMessage == nil {
                    errorMessage = item.error?.localizedDescription
                }
            }
            isBuffering = player.timeControlStatus == .waitingToPlayAtSpecifiedRate
            isPlaying = player.timeControlStatus == .playing
        case .idle:
            break
        }
    }

    func step() {
        guard mode == .slideshow, currentIndex + 1 < segments.count else { return }
        currentIndex += 1
        elapsed = 0
    }

    func back() {
        guard mode == .slideshow, currentIndex > 0 else { return }
        currentIndex -= 1
        elapsed = 0
    }

    func seek(toFraction fraction: Double) {
        let fraction = min(max(fraction, 0), 1)
        switch mode {
        case .slideshow:
            guard !segments.isEmpty else { return }
            var remaining = fraction * totalDuration
            var index = segments.count - 1
            for (i, segment) in segments.enumerated() {
                if remaining <= segment.duration {
                    index = i
                    break
                }
                remaining -= segment.duration
            }
            currentIndex = min(index, segments.count - 1)
            elapsed = min(remaining, segments[currentIndex].duration)
        case .avplayer:
            guard totalDuration > 0, let player else { return }
            playerTime = fraction * totalDuration
            player.seek(
                to: CMTime(seconds: playerTime, preferredTimescale: 600),
                toleranceBefore: .zero,
                toleranceAfter: .zero
            )
        case .idle:
            break
        }
    }

    func seek(to time: Double) {
        guard totalDuration > 0 else { return }
        seek(toFraction: time / totalDuration)
    }

    // MARK: - 生命周期

    func startTicking() {
        guard tickTask == nil else { return }
        tickTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: .milliseconds(50))
                self?.tick(delta: 0.05)
            }
        }
    }

    func stopTicking() {
        tickTask?.cancel()
        tickTask = nil
    }

    /// 彻底放手：停播放器、卸载本地挂载、清缓存。
    func stop() {
        stopTicking()
        teardown()
    }

    private func teardown() {
        player?.pause()
        player?.replaceCurrentItem(with: nil)
        player = nil
        if let source {
            HLSLocalServer.shared.unregister(source)
        }
        source = nil
        reset()
    }

    private func reset() {
        mode = .idle
        segments = []
        currentIndex = 0
        elapsed = 0
        isPlaying = false
        isBuffering = false
        playerTime = 0
        playerDuration = 0
    }

    private func applyRate() {
        guard mode == .avplayer, let player else { return }
        player.rate = isPlaying ? Float(max(speed, 0.25)) : 0
    }

    private func configureAudioSession() {
        #if os(iOS)
        let session = AVAudioSession.sharedInstance()
        try? session.setCategory(.playback, mode: .moviePlayback)
        try? session.setActive(true)
        #endif
    }
}
